File: web_frontend/utils/msgModel.js
// 定义常量
const MsgStatusSent = "sent";         // 消息已发送
const MsgStatusDelivered = "delivered"; // 消息已送达
const MsgStatusRead = "read";         // 消息已读

const MsgTypeUser = "private";       // 用户消息
const MsgTypeGroup = "group";        // 群组消息
const MsgTypeAck = "ack";            // Ack 消息
const MsgTypeSystem = "system";      // 系统消息

const MsgContentTypeText = "text";   // 消息类型：文本消息
const MsgContentTypeImage = "image"; // 消息类型：图片消息
class WsMsg {
    constructor(msgType, status, wsHeader, wsBody) {
        this.msg_id = generateRandomID(); // 自动生成唯一的消息 ID
        this.msg_type = msgType;
        this.status = status;
        this.wsHeader = wsHeader;
        this.wsBody = wsBody;
    }
}

class WsHeader {
    constructor(senderId, receiverId = [], msgContentType, timestamp, groupId = null) {
        this.sender_id = senderId;
        this.receiver_id = receiverId; // 这里是数组，包含所有接收者的 ID
        this.msg_content_type = msgContentType;
        this.timestamp = timestamp;
        this.group_id = groupId; // 只有群消息时才会设置
    }
}
class WsBody {
    constructor(msgContent) {
        this.msg_content = msgContent;
    }
}

function newPrivateMessage(senderId, msgContentType, content,receiverId=[]) {
    return new WsMsg(
        MsgTypeUser,
        MsgStatusSent,
        new WsHeader(
            senderId,
            receiverId,
            msgContentType,
            Date.now()
        ),
        new WsBody(content)
    );
}
function newGroupMessage(senderId, groupId, msgContentType, content, receiverId=[]) {
    return new WsMsg(
        MsgTypeGroup,
        MsgStatusSent,
        new WsHeader(
            senderId,
            receiverId,
            msgContentType,
            Date.now(),
            groupId
        ),
        new WsBody(content)
    );
}
// flatten 函数


// 生成随机 ID 的函数
function generateRandomID() {
    // 生成 16 字节的随机数
    const bytes = crypto.getRandomValues(new Uint8Array(16));
    // 将字节数组转换为十六进制字符串
    return Array.from(bytes).map(byte => byte.toString(16).padStart(2, '0')).join('');
}
function newTestPrivateMessage(senderId, receiverId, content) {
    let msg = {
        //随机生成消息ID
        "msg_id": generateRandomID(),
        "msg_type": "private",
        "status": "sent",
        "sender_id": senderId.toString(),
        "receiver_id": [
            receiverId
        ],
        "msg_content_type": "text",
        "timestamp": Date.now(),
        "msg_content": content
    }
    return msg;


}
//receiverIds为数组
function newTestGroupMessage(senderId, groupId, receiverIds, content) {
    
    let msg ={
        "msg_id": generateRandomID(),
        "msg_type": "group",
        "status": "sent",
        "sender_id": senderId.toString(),
        "receiver_id": receiverIds, // 直接使用传递的数组
        "msg_content_type": "text",
        "timestamp": Date.now(),
        "group_id": groupId.toString(),
        "msg_content": content
      }
      return msg;


}
// 导出函数
export { newTestPrivateMessage, newTestGroupMessage };

File: ws_server/shared/MsgModel.go
package shared

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"
)

const (
	MsgStatusSent      = "sent"      // 消息已发送
	MsgStatusDelivered = "delivered" // 消息已送达
	MsgStatusRead      = "read"      // 消息已读
)
const (
	// 用户消息
	MsgTypeUser = "private"
	// // 群组消息
	MsgTypeGroup = "group"
	// Ack 消息
	MsgTypeAck = "ack"
	// 系统消息
	MsgTypeSystem = "system"
	// 邀请消息
	MsgTypeInvition = "invition"
)
const (
	// 消息类型：文本消息
	MsgContentTypeText = "text"
	// 消息类型：图片消息
	MsgContentTypeImage = "image"
)

type WsMsg struct {
	MsgID   string `json:"msg_id"`   // 消息唯一标识符
	MsgType string `json:"msg_type"` // 消息类型(用户消息/群组消息/Ack 消息/系统消息/邀请消息)
	Status  string `json:"status"`   // 消息状态 (sent/delivered/read)
	WsHeader
	WsBody
}

type WsHeader struct {
	SenderId       string   `json:"sender_id"`
	ReceiverId     []string `json:"receiver_id"`
	MsgContentType string   `json:"msg_content_type"`   // 消息内容类型 (text/image)
	Timestamp      int64    `json:"timestamp"`          // 时间戳
	GroupId        string   `json:"group_id",omitempty` // 群组ID（可选）
}
type WsBody struct {
	MsgContent string                 `json:"msg_content"`
	Extra      map[string]interface{} `json:"extra,omitempty"` // 扩展字段，用于自定义场景
}

// 生成随机 ID
func generateRandomID() string {
	bytes := make([]byte, 16) // 生成 16 字节的随机数
	_, err := rand.Read(bytes)
	if err != nil {
		return ""
	}
	return hex.EncodeToString(bytes)
}

// 创建一个新的消息
func NewWsUserMessage(senderId string, receiverId []string, msgContent string) WsMsg {
	header := WsHeader{
		SenderId:       senderId,
		ReceiverId:     receiverId,
		MsgContentType: MsgContentTypeText,
		Timestamp:      time.Now().UnixNano(),
	}

	body := WsBody{
		MsgContent: msgContent,
	}

	return WsMsg{
		MsgID:    generateRandomID(),
		MsgType:  MsgTypeUser,
		Status:   MsgStatusSent,
		WsHeader: header,
		WsBody:   body,
	}
}

func NewWsGroupMessage(receiverId []string, senderId, msgContent, GroupId string) WsMsg {
	header := WsHeader{
		SenderId:       senderId,
		ReceiverId:     receiverId,
		MsgContentType: MsgContentTypeText,
		Timestamp:      time.Now().UnixNano(),
		GroupId:        GroupId,
	}

	body := WsBody{
		MsgContent: msgContent,
	}

	return WsMsg{
		MsgID:    generateRandomID(),
		MsgType:  MsgTypeGroup,
		Status:   MsgStatusSent,
		WsHeader: header,
		WsBody:   body,
	}
}

// 消息格式美化
func (msg *WsMsg) PrettyPrint() string {
	formattedJSON, err := json.MarshalIndent(msg, "", "  ")
	if err != nil {
		return fmt.Sprintf("Failed to format message: %v", err)
	}
	return string(formattedJSON)
}
